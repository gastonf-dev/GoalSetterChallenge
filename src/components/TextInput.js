/**
 * @flow stric-local
 * @format
 */
import React, {useState} from 'react';
import type {Node} from 'react';
import {
  StyleSheet,
  Text,
  View,
  TextInput as TextInputRN,
  Image,
} from 'react-native';
import COLORS from '../styles/Colors';

const ICONS = {
  person: require('../assets/images/person.png'),
  email: require('../assets/images/emailIcon.png'),
  password: require('../assets/images/passwordIcon.png'),
};

type Props = {
  label: string,
  placeholder: string,
  type: 'person' | 'email' | 'password',
  value: string,
  handleChange: function,
  handleBlur: function,
  error: string,
  touched: boolean,
  name: string,
};

const TextInput = ({
  label,
  placeholder,
  type,
  handleBlur,
  handleChange,
  value,
  name,
  error,
  touched,
}: Props): Node => {
  const [isFocused, setIsFocused] = useState(false);

  return (
    <View
      style={[
        styles.container,
        isFocused ? {borderBottomColor: COLORS.lightblue} : null,
        (type == 'email' || type == 'password') && error && touched
          ? {borderBottomColor: COLORS.red}
          : null,
      ]}>
      <Image source={ICONS[type]} style={styles.icon} />
      <View style={styles.textInputContainer}>
        {isFocused || value.length !== 0 ? (
          <Text style={styles.label}>{label}</Text>
        ) : null}
        <TextInputRN
          style={styles.textInput}
          placeholder={isFocused ? undefined : placeholder}
          onFocus={() => setIsFocused(true)}
          onBlur={event => {
            setIsFocused(false);
            handleBlur(event);
          }}
          secureTextEntry={type === 'password'}
          onChangeText={handleChange}
          value={value}
        />
      </View>
    </View>
  );
};

export default TextInput;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    borderBottomWidth: 1,
    borderBottomColor: COLORS.gray,
    marginHorizontal: 38,
    marginVertical: 12,

    paddingVertical: 6,
    // Apply fontFamily to all nest components
    fontFamily: 'AvenirLTStd',
    fontWeight: 'bold',
  },
  textInputContainer: {
    flexDirection: 'column',
    justifyContent: 'flex-start',
  },
  textInput: {
    fontSize: 14,
    color: COLORS.lightblue,
    width: '100%',
  },
  label: {
    fontSize: 10,
    color: COLORS.gray,
    marginVertical: 6
  },
  icon: {
    marginRight: 12,
    tintColor: COLORS.gray,
  },
});
