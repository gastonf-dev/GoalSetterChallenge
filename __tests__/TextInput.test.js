import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent, waitFor} from '@testing-library/react-native';

import TextInput from '../src/components/TextInput';
describe('<TextInput />', () => {
  test('render input without value', () => {
    const WITHOUT_VALUE = {
      label: 'Test label',
      placeholder: 'Test placeholder',
      type: 'person',
      handleBlur: jest.fn(),
      handleChange: jest.fn(),
      value: '',
      name: 'test',
      error: '',
      touched: false,
    };
    const wrapper = renderer.create(<TextInput {...WITHOUT_VALUE} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('render input without value', () => {
    const WITH_VALUE = {
      label: 'Test label',
      placeholder: 'Test placeholder',
      type: 'person',
      handleBlur: jest.fn(),
      handleChange: jest.fn(),
      value: 'Test value',
      name: 'test',
      error: '',
      touched: false,
    };
    const wrapper = renderer.create(<TextInput {...WITH_VALUE} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('should exists label component when render input with value', () => {
    const WITH_VALUE = {
      label: 'Test label',
      placeholder: 'Test placeholder',
      type: 'person',
      handleBlur: jest.fn(),
      handleChange: jest.fn(),
      value: 'Test value',
      name: 'test',
      error: '',
      touched: false,
    };
    const {getByTestId} = render(<TextInput {...WITH_VALUE} />);

    expect(getByTestId('label')).toBeTruthy();
  });
});
