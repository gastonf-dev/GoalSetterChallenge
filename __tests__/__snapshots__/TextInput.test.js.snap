// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> render input without value 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderBottomColor": "#D8D8D8",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "marginHorizontal": 38,
        "marginVertical": 12,
        "paddingVertical": 12,
      },
      null,
      null,
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/person.png",
      }
    }
    style={
      Object {
        "marginRight": 12,
        "tintColor": "#D8D8D8",
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="words"
      keyboardType="default"
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onFocus={[Function]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Object {
          "color": "#17BFDD",
          "fontFamily": "Avenir LT Std",
          "fontSize": 14,
          "fontWeight": "bold",
          "lineHeight": 16,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;

exports[`<TextInput /> render input without value 2`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderBottomColor": "#D8D8D8",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "marginHorizontal": 38,
        "marginVertical": 12,
        "paddingVertical": 12,
      },
      null,
      null,
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/person.png",
      }
    }
    style={
      Object {
        "marginRight": 12,
        "tintColor": "#D8D8D8",
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#D8D8D8",
          "fontSize": 10,
          "lineHeight": 12,
          "marginVertical": 6,
        }
      }
      testID="label"
    >
      Test label
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="words"
      keyboardType="default"
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onFocus={[Function]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Object {
          "color": "#17BFDD",
          "fontFamily": "Avenir LT Std",
          "fontSize": 14,
          "fontWeight": "bold",
          "lineHeight": 16,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
      value="Test value"
    />
  </View>
</View>
`;
